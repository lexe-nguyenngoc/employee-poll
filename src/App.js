import React from "react";
import { useRoutes } from "react-router-dom";
import NotFound from "./components/NotFound";
import Auth from "./features/auth";
import EmployeePoll from "./features/employee-poll";
import LoginPage from "./features/auth/pages/Login";
import SignUpPage from "./features/auth/pages/SignUp";

const App = () => {
  const routes = useRoutes([
    {
      path: "/",
      element: <EmployeePoll />,
      children: []
    },
    {
      path: "auth",
      element: <Auth />,
      children: [
        {
          path: "",
          element: <LoginPage />
        },
        {
          path: "signup",
          element: <SignUpPage />
        }
      ]
    },
    {
      path: "*",
      element: <NotFound />
    }
  ]);

  return routes;
};

export default App;
