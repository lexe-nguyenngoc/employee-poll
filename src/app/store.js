import { configureStore } from "@reduxjs/toolkit";
import counterReducer from "../features/counter/counterSlice";
import authReducer, {
  ROOT_STATE_NAME as auth
} from "../features/auth/authSlice";
import employeePollReducer, {
  ROOT_STATE_NAME as employeePollName
} from "../features/employee-poll/employeePollSlice";

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    [auth]: authReducer,
    [employeePollName]: employeePollReducer
  }
});
