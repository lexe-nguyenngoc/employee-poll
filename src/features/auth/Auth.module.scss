.auth {
  width: 600px;
  margin: 0 auto;
  padding: 16px;

  &__image {
    width: 60%;
    margin: 0 auto;
    display: block;
  }
}
