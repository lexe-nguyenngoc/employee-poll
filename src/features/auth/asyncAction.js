import { createAsyncThunk } from "@reduxjs/toolkit";
import { _getUsers } from "../../api";

export const loginAsync = createAsyncThunk(
  "auth/login",
  async (user, thunkAPI) => {
    const users = await _getUsers();
    const foundUser = users[user.user];
    if (foundUser?.password === user.password) return foundUser;
    return thunkAPI.rejectWithValue("This user is incorrect!");
  }
);
