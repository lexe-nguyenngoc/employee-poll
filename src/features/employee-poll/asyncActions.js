import { createAsyncThunk } from "@reduxjs/toolkit";
import { _getQuestions, _getUsers } from "../../api";

export const getQuestionsThunk = createAsyncThunk(
  "employeePoll/getQuestions",
  async (_) => {
    const questions = await _getQuestions();

    return questions;
  }
);

export const getQuestionThunk = createAsyncThunk(
  "employeePoll/getQuestionDetail",
  async ({ id, failFn }) => {
    const questions = await _getQuestions();
    if (questions[id]) return questions[id];

    failFn();
  }
);

export const getUsersThunk = createAsyncThunk(
  "employeePoll/getUsers",
  async (_) => {
    const users = await _getUsers();
    return users;
  }
);
