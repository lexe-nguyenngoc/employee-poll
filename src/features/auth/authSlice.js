import { createSlice } from "@reduxjs/toolkit";
import { loginAsync } from "./asyncActions";
import { saveQuestionAnswer } from "../employee-poll/asyncActions";

export const ROOT_STATE_NAME = "auth";

const authSlice = createSlice({
  name: "auth",
  initialState: {
    loading: false,
    errorMessage: null,
    currentUser: null
  },
  reducers: {
    resetErrorMessage: (state) => {
      state.errorMessage = null;
    },
    logout: (state) => {
      state.currentUser = null;
    },
    changeAnswer: (state, action) => {
      const { qid, answer } = action.payload;
      state.currentUser.answers[qid] = answer;
    }
  },
  extraReducers: (builder) => {
    builder
      .addCase(loginAsync.pending, (state) => {
        state.loading = true;
        state.errorMessage = null;
      })
      .addCase(loginAsync.fulfilled, (state, action) => {
        state.loading = false;
        state.currentUser = action.payload;
        state.errorMessage = null;
      })
      .addCase(loginAsync.rejected, (state, action) => {
        state.loading = false;
        state.errorMessage = action.payload;
      })
      .addCase(saveQuestionAnswer.fulfilled, (state, action) => {
        const { qid, answer } = action.payload.questionAnswer;
        state.currentUser.answers[qid] = answer;
      });
  }
});

export const { resetErrorMessage, logout, changeAnswer } = authSlice.actions;
export const selectState = (state) => state[ROOT_STATE_NAME];
export const selectCurrentUser = (state) => state[ROOT_STATE_NAME].currentUser;
export default authSlice.reducer;
