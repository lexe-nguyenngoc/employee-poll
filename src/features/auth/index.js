import React from "react";
import { Navigate, Outlet } from "react-router-dom";
import classNames from "classnames/bind";

import authImage from "../../assets/images/login.png";

import styles from "./Auth.module.scss";
import { useSelector } from "react-redux";
import { selectCurrentUser } from "./authSlice";

const cx = classNames.bind(styles);

const Auth = () => {
  const currentUser = useSelector(selectCurrentUser);

  if (currentUser) return <Navigate to="/" />;

  return (
    <div className={cx("auth")}>
      <img className={cx("auth__image")} src={authImage} alt="Auth" />

      <Outlet />
    </div>
  );
};

export default Auth;
