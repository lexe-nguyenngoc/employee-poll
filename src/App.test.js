import { fireEvent, render } from "@testing-library/react";
import * as React from "react";
import App from "./App";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "./app/store";
import { loginAsync } from "./features/auth/asyncActions";

describe("App component", () => {
  test("should match with snapshot", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });

  test("should show login page when user is not login", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(view.queryByText(/login/i)).toBeInTheDocument();
  });

  test("should show dashboard page when user is login", async () => {
    await store.dispatch(
      loginAsync({ user: "sarahedo", password: "password123" })
    );

    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    expect(view.queryByText(/home/i)).toBeInTheDocument();
  });

  test("should show leaderboard when user is login", async () => {
    await store.dispatch(
      loginAsync({ user: "sarahedo", password: "password123" })
    );

    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    const link = view.queryByText(/leaderboard/i);
    expect(link).toBeInTheDocument();
    fireEvent.click(link);

    expect(view.queryByText(/users/i)).toBeInTheDocument();
    expect(view.queryByText(/answered/i)).toBeInTheDocument();
    expect(view.queryByText(/created/i)).toBeInTheDocument();
  });

  test("should show new page when user is login", async () => {
    await store.dispatch(
      loginAsync({ user: "sarahedo", password: "password123" })
    );

    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    const link = view.queryByText(/new/i);
    expect(link).toBeInTheDocument();
    fireEvent.click(link);

    expect(view.queryByText(/would you rather/i)).toBeInTheDocument();
    expect(view.queryByText(/create your own poll/i)).toBeInTheDocument();
    expect(view.queryByText(/first option/i)).toBeInTheDocument();
    expect(view.queryByText(/second option/i)).toBeInTheDocument();
    expect(view.queryByText(/submit/i)).toBeInTheDocument();
  });

  test("should show login page when user is login and click logout", async () => {
    await store.dispatch(
      loginAsync({ user: "sarahedo", password: "password123" })
    );

    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    const btn = view.queryByText(/logout/i);
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);

    expect(view.queryByText(/login/i)).toBeInTheDocument();
  });
});
