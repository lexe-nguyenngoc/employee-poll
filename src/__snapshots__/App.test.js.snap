// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="loading"
      />
      <div
        class="auth"
      >
        <img
          alt="Auth"
          class="auth__image"
          src="login.png"
        />
        <form
          class="form"
        >
          <h3
            class="form__heading"
          >
            Login
          </h3>
          <div
            class="form__container"
          >
            <div
              class="form-field"
            >
              <label
                class="form-field__label required"
                for=":r0:"
              >
                User
              </label>
              <div
                class="form-field__control"
              >
                <input
                  class="input-field"
                  id=":r0:"
                  name="user"
                  placeholder="Input your user"
                  type="text"
                />
              </div>
              <div
                class="form-field__message"
              />
            </div>
            <div
              class="form-field"
            >
              <label
                class="form-field__label required"
                for=":r1:"
              >
                Password
              </label>
              <div
                class="form-field__control"
              >
                <input
                  class="input-field"
                  id=":r1:"
                  name="password"
                  placeholder="Input your password"
                  type="password"
                />
              </div>
              <div
                class="form-field__message"
              />
            </div>
            <button
              class="btn contained primary"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="loading"
    />
    <div
      class="auth"
    >
      <img
        alt="Auth"
        class="auth__image"
        src="login.png"
      />
      <form
        class="form"
      >
        <h3
          class="form__heading"
        >
          Login
        </h3>
        <div
          class="form__container"
        >
          <div
            class="form-field"
          >
            <label
              class="form-field__label required"
              for=":r0:"
            >
              User
            </label>
            <div
              class="form-field__control"
            >
              <input
                class="input-field"
                id=":r0:"
                name="user"
                placeholder="Input your user"
                type="text"
              />
            </div>
            <div
              class="form-field__message"
            />
          </div>
          <div
            class="form-field"
          >
            <label
              class="form-field__label required"
              for=":r1:"
            >
              Password
            </label>
            <div
              class="form-field__control"
            >
              <input
                class="input-field"
                id=":r1:"
                name="password"
                placeholder="Input your password"
                type="password"
              />
            </div>
            <div
              class="form-field__message"
            />
          </div>
          <button
            class="btn contained primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App component should show login page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="loading"
      />
      <div
        class="auth"
      >
        <img
          alt="Auth"
          class="auth__image"
          src="login.png"
        />
        <form
          class="form"
        >
          <h3
            class="form__heading"
          >
            Login
          </h3>
          <div
            class="form__container"
          >
            <div
              class="form-field"
            >
              <label
                class="form-field__label required"
                for=":r2:"
              >
                User
              </label>
              <div
                class="form-field__control"
              >
                <input
                  class="input-field"
                  id=":r2:"
                  name="user"
                  placeholder="Input your user"
                  type="text"
                />
              </div>
              <div
                class="form-field__message"
              />
            </div>
            <div
              class="form-field"
            >
              <label
                class="form-field__label required"
                for=":r3:"
              >
                Password
              </label>
              <div
                class="form-field__control"
              >
                <input
                  class="input-field"
                  id=":r3:"
                  name="password"
                  placeholder="Input your password"
                  type="password"
                />
              </div>
              <div
                class="form-field__message"
              />
            </div>
            <button
              class="btn contained primary"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="loading"
    />
    <div
      class="auth"
    >
      <img
        alt="Auth"
        class="auth__image"
        src="login.png"
      />
      <form
        class="form"
      >
        <h3
          class="form__heading"
        >
          Login
        </h3>
        <div
          class="form__container"
        >
          <div
            class="form-field"
          >
            <label
              class="form-field__label required"
              for=":r2:"
            >
              User
            </label>
            <div
              class="form-field__control"
            >
              <input
                class="input-field"
                id=":r2:"
                name="user"
                placeholder="Input your user"
                type="text"
              />
            </div>
            <div
              class="form-field__message"
            />
          </div>
          <div
            class="form-field"
          >
            <label
              class="form-field__label required"
              for=":r3:"
            >
              Password
            </label>
            <div
              class="form-field__control"
            >
              <input
                class="input-field"
                id=":r3:"
                name="password"
                placeholder="Input your password"
                type="password"
              />
            </div>
            <div
              class="form-field__message"
            />
          </div>
          <button
            class="btn contained primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
